File: backend/app/models/transcription.py
from sqlalchemy import Column, String, Integer, Float, Boolean, DateTime, Text, JSON, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.db.base_class import Base

class Transcription(Base):
    __tablename__ = "transcriptions"
    
    id = Column(String(36), primary_key=True)
    session_id = Column(String(36), ForeignKey("counseling_sessions.id"), nullable=False)
    
    # Full transcription text
    full_text = Column(Text, nullable=False)
    
    # Language and metadata
    language = Column(String(10), default="ja")
    duration = Column(Float, nullable=True)
    
    # Processing status
    status = Column(String(20), default="pending")
    # pending, processing, completed, failed
    
    # Segments with timestamps and speaker information
    segments = Column(JSON, nullable=True)
    
    # Speaker statistics
    speaker_stats = Column(JSON, nullable=True)
    
    # Processing metadata
    processing_time = Column(Float, nullable=True)  # seconds
    error_message = Column(Text, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationship
    session = relationship("CounselingSession", back_populates="transcription")

class TranscriptionSegment(Base):
    __tablename__ = "transcription_segments"
    
    id = Column(String(36), primary_key=True)
    transcription_id = Column(String(36), ForeignKey("transcriptions.id"), nullable=False)
    
    # Segment details
    segment_index = Column(Integer, nullable=False)
    start_time = Column(Float, nullable=False)
    end_time = Column(Float, nullable=False)
    text = Column(Text, nullable=False)
    
    # Speaker information
    speaker = Column(String(20), nullable=True)  # counselor, client
    speaker_confidence = Column(Float, nullable=True)
    
    # Editing
    is_edited = Column(Boolean, default=False)
    original_text = Column(Text, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationship
    transcription = relationship("Transcription")

File: frontend/app/repository/transcription.ts
import { apiRequest } from './base';

export interface TranscriptionSegment {
  id: number;
  start: number;
  end: number;
  text: string;
  speaker: 'counselor' | 'client';
  speaker_confidence?: number;
  is_edited?: boolean;
  original_text?: string;
}

export interface SpeakerStats {
  counselor: {
    total_time: number;
    segment_count: number;
    percentage: number;
  };
  client: {
    total_time: number;
    segment_count: number;
    percentage: number;
  };
}

export interface TranscriptionData {
  transcription_id: string;
  session_id: string;
  full_text: string;
  language: string;
  duration: number;
  segments: TranscriptionSegment[];
  speaker_stats: SpeakerStats;
  processing_time: number;
  created_at: string;
  updated_at: string;
}

export interface TranscriptionStatus {
  session_id: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  transcription_id?: string;
  processing_time?: number;
  language?: string;
  duration?: number;
  speaker_stats?: SpeakerStats;
}

export interface TaskStatus {
  task_id: string;
  state: 'PENDING' | 'PROCESSING' | 'SUCCESS' | 'FAILURE';
  progress: number;
  stage: string;
  result?: any;
  error?: string;
  traceback?: string;
}

export async function startTranscription(sessionId: string): Promise<{ task_id: string }> {
  return apiRequest<{ task_id: string }>(`/transcriptions/${sessionId}/start`, {
    method: 'POST',
  });
}

export async function getTranscriptionStatus(sessionId: string): Promise<TranscriptionStatus> {
  return apiRequest<TranscriptionStatus>(`/transcriptions/${sessionId}/status`);
}

export async function getTranscription(sessionId: string): Promise<TranscriptionData> {
  return apiRequest<TranscriptionData>(`/transcriptions/${sessionId}`);
}

export async function updateTranscriptionSegment(
  transcriptionId: string,
  segmentIndex: number,
  newText: string
): Promise<void> {
  await apiRequest(`/transcriptions/${transcriptionId}/segments/${segmentIndex}`, {
    method: 'PATCH',
    body: JSON.stringify({ new_text: newText }),
  });
}

export async function getTaskStatus(taskId: string): Promise<TaskStatus> {
  return apiRequest<TaskStatus>(`/tasks/${taskId}/status`);
}